### DISCUSSION TITLE
Is there a way to have states live in directories other than the directory of the base abstract class?

### DISCUSSION THREAD

**Discussion by javoscript** [spatie/laravel-model-states#202] [Q&A] [answered]
I'm trying to generate an extendable structure where I provide a set of "base" states for a given field, but I want to be able to later add more states (from other directories). This would allow, for example, to provide extendable packages that provide a set of already implemented states.

Haven´t found any way of manually registering or adding extra available states for a given field.
And in the [resolveStateMapping](https://github.com/spatie/laravel-model-states/blob/ebdfb658a8990dc0d24f2f51670b3c1a09c7c539/src/State.php#L274) method, it only looks for states in the same directory as the base abstract state class.

In the docs, it only mentions that the state classes should live in the same base state class directory if you want to use custom names ([docs](https://spatie.be/docs/laravel-model-states/v2/working-with-states/02-serializing-states#content-resolving-states-from-the-database)).

But if I don't locate the concrete state classes in the same directory, I get some errors whenever the package wants to cast the class name string value to the corresponding state.

Say I define the base abstract state class at `app/States/ExampleModel/ExampleState.php`. If I then implement a state at `app/States/ExampleModel/Base/ExampleStateImplementation.php` and extend the base abstract class, whenever I want to save the model with this state, I get this error:

```
<warning>PHP Warning:  Undefined array key "App\States\ExampleModel\Base\ExampleStateImplementation" in /var/www/html/vendor/laravel/framework/src/Illuminate/Collections/Collection.php on line 1683</warning>
PHP Error:  Class name must be a valid object or a string in /var/www/html/vendor/spatie/laravel-model-states/src/StateCaster.php on line 30
```

If I try to get the state mapping from the base state class, I get an empty collection:

```
>>> \App\States\ExampleModel\ExampleState::getStateMapping()
=> Illuminate\Support\Collection {#3503
     all: [],
   }
```

**Comment by brendt** [comment]
Yeah that's not supported, concrete state implementations must live in the same directory as the abstract class.

**Reply by javoscript** [reply]
Just published a pull request with a potential implementation at #203
Please let me know if there is anything you'd like be implemented differently :smile:

**Reply by MrJuliuss** [reply]
Hi, i need this feature too :)

**Comment by javoscript** [comment, marked as answer]
Merged in 2.4.0
